Fix/7563 improving admin search component

zmaglica opened the pull request:
Fixes woocommerce/woocommerce#32239 

The search component has UX issues. When multiple tags are added, the component breaks apart. To fix this issue, the flex property is added to the parent element and gradient is added by using CSS after property to simulate diapering effect when tags can't be placed inside the search component.  The default property of shrinking column has been disabled and min width has been added to multiple elements to improve UI experience to the end user. 

### Screenshots

**Before**:

![image](https://user-images.githubusercontent.com/15913573/143605502-1ceb9b09-bfae-449e-8dff-a0c882ae870c.png)

<img width="421" alt="Screen Shot 2021-08-25 at 8 14 43 PM" src="https://user-images.githubusercontent.com/4500952/130895561-e2b845a1-e654-4e8d-9e11-8dc8dadc4cf2.png">

**After**:
Desktop screen when content can fit:

![image](https://user-images.githubusercontent.com/15913573/143602697-84b74448-4cf1-4159-97db-79db4e793b1c.png)

Mobile screen when content can't fit:

![image](https://user-images.githubusercontent.com/15913573/143603025-d25c87d5-c40a-4092-ada4-83fd88eaae4d.png)


### Detailed test instructions:

-   Open Customers page in WooCommerce or any page that uses this component
-   Enter multiple tags in search box
-   The input should handle multiple tags and scale them properly

Or take a look at testing instructions [here](woocommerce/woocommerce#32239).

zmaglica committed:
Improved UI/UX search componenet

zmaglica committed:
Added missing javscript file

zmaglica committed:
Fixed issue when no tags are entered

RadoslavGeorgiev reviewed (changes requested):
This is a nice start! We need to figure out a few details first though :)

## Overlap

There is "scroll" within the tags container, but it gets a bit quirky when trying to add a new tag, since it is outside the scrollable zone:

<img width="194" alt="overlap-2" src="https://user-images.githubusercontent.com/5311119/143868556-1b241160-9001-4188-86c2-f43ff4d8c061.png">

## Linting issues

There seem to be some linting issues, which need to be fixed. Have you managed to get your editor to display them?

zmaglica committed:
Added gradient on overflow

zmaglica committed:
Code review fixes:
- Added gradient on overflow
- Removed JavaScript code that was used before to detect overflow
- Added minimum width for mobile and desktop screens for search input element
- Added search input field inside tags element to maintain better UX on small screen

zmaglica commented on packages/components/src/select-control/tags.js:
```diff
@@ -39,15 +51,20 @@ class Tags extends Component {
 		return (
 			<Fragment>
-				<div className="woocommerce-select-control__tags">
+				<div ref={ ref => this.tagsParent = ref } className={`woocommerce-select-control__tags${overflowActive ? " is-overflown" : ""}`}>
```
This part of code is refactored. Overflown is no longer calculated and ref is no longer needed at this point, so it was removed.

zmaglica commented on packages/components/src/select-control/tags.js:
```diff
@@ -21,6 +21,18 @@ class Tags extends Component {
+		};
+	}
+
+	isElementOverflown(e) {
```
This piece of code is removed since it no longer needed/used. CSS class that was used to add overflown properties was added to the original div element.

zmaglica commented on packages/components/src/select-control/tags.js:
```diff
@@ -21,6 +21,18 @@ class Tags extends Component {
+		if (typeof e === 'undefined') {
+			return false;
+		}
+		const totalWidth = Object.values(e.childNodes).reduce((total, i) => total + i.clientWidth, 0);
```
This function is removed. Check https://github.com/woocommerce/woocommerce-admin/pull/7972#discussion_r759351730

zmaglica commented on packages/components/src/select-control/tags.js:
```diff
@@ -39,15 +51,20 @@ class Tags extends Component {
 	}
 
+	componentDidMount() {
+		this.setState({ overflowActive: this.isElementOverflown(this.tagsParent) });
```
State is no longer needed because logic behind detecting overflown elements is refactored. Check previous comments.

zmaglica commented on packages/components/src/search/style.scss:
```diff
@@ -62,9 +62,20 @@
+		max-width: calc(100% - 60px);
+	}
+
+	.is-overflown {
```
Point taken. The CSS code in this class is added to the element and this class is deleted.

RadoslavGeorgiev reviewed (changes requested):
The PR is shaping up really well 👏 🥳 

I left a couple of minor comments within the code, and noticed a couple of additional things:

Whenever there are no tags or few of them, the text field does not take up all of the available width, causing the text inside of it to truncate.

<img width="654" alt="image" src="https://user-images.githubusercontent.com/5311119/144102463-40c3ee9d-eb6a-401b-8732-9b32b8c9bf89.png">

Seems like the text field is growing within the list of tags, but the list is only taking up a part of the available space.

Adding this seems to fix it:

```css
`.woocommerce-search.woocommerce-select-control .components-base-control .woocommerce-select-control__tags {
    width: 100%;
}
```

There are some layout issues, which were not present before, as the text was wrapping:

<img width="818" alt="image" src="https://user-images.githubusercontent.com/5311119/144102670-993e0d44-fad5-42fe-8567-0fb139fd771e.png">

Adding those styles seems to fix the issue:

```
.woocommerce-table .woocommerce-table__actions > div {
    overflow: hidden;
}
```
Note that `overflow: hidden` also clears floats and whatnot, whenever there are no elements with position: absolute.

zmaglica commented on packages/components/src/select-control/control.js:
```diff
@@ -96,21 +96,6 @@ class Control extends Component {
 		}
 	}
-
-	renderButton() {
```
My IDE reported that this function is no longer used. I even checked all child components and I couldn't find usage of this function. Yet, pipelines reported that one E2E test failed (`Enabling cash on delivery enables the payment method`), but I can't see why it failed because AFAIK I haven't touched that part of the plugin. I can return this function and execute E2E tests again and check do they pass.

zmaglica commented:
> Adding those styles seems to fix the issue:

```
.woocommerce-table .woocommerce-table__actions > div {
    overflow: hidden;
}
```
Adding this will prevent appearance of suggestion box (_"All customers that starts with ..."_) .I will try to fix this using different approach.

zmaglica committed:
Code review fixes:
- Removed unnecessary flex-shrink property.
- Removed browser specific calc properties.
- Replaced brackets with quotes on classname where classnames are not dynamically created
- Fixed issue when search term is truncated when search term is long

zmaglica committed:
Code review fixes:
- Fixed issue on table when download action is added

zmaglica committed:
- Refactored CSS code for table to fix issues when actions are added
- Tweaked CSS code for select control and search components
- Minor improvements for Control component

RadoslavGeorgiev reviewed (approved):
The PR looks and works great, thank you @zmaglica! 🥳 🚢 

Please make sure to add a changelog entry before merging :)

zmaglica committed:
Added changelog

zmaglica committed:
Fixed typo in changelog

joshuatf commented on packages/components/src/select-control/control.js:
```diff
@@ -227,38 +246,11 @@ class Control extends Component {
-						</p>
-					) }
-				</div>
+				{ inlineTags ? (
```
This logic (moving field separately) isn't quite correct.  With this logic, `inlineTags` will never get hit inside of `field` since it's not reached.

joshuatf commented on packages/components/src/table/style.scss:
```diff
@@ -15,7 +15,7 @@
 
 		> div {
-			width: 100%;
+			flex-grow: 1;
```
Nice!

joshuatf commented on packages/components/src/select-control/tags.js:
```diff
@@ -68,17 +67,15 @@ class Tags extends Component {
+						isLink={ false }
 						onClick={ this.removeAll }
 					>
-						<Icon
```
I think this looks correct based on the original issue and the fact that tags also have an "x", but just to verify the intention is to remove the "x" icon for the clear button, right @LevinMedia?  cc @elizaan36

joshuatf reviewed (changes requested):
Thanks for the UX fixes, @zmaglica!  Things are looking pretty good here, but a couple minor issues remain.

If you run `npm run storybook` you can also take a look at all use cases simultaneously for `SelectControl`.  Right now I see this one as broken:

![Screen Shot 2021-12-17 at 10 01 45 AM](https://user-images.githubusercontent.com/10561050/146565142-ef7255a7-b16b-4fd6-940e-e9933e9f0a3f.png)

joshuatf commented:
Hi @zmaglica 👋  just wanted to check in and see if you're still working on this issue.

adrianduffell commented:
Hi @zmaglica, just checking if you are still working on this?

We're planning to [merge Woo Admin to Core](https://developer.woocommerce.com/2021/10/14/rfc-woocommerce-admin-is-merging-into-core/) soon, and ideally it would be best have this PR completed before then.

chihsuan commented:
Hi @zmaglica We've merged Woo Admin into core. This is a nice PR!  Would you be able to recreate a PR in [woocommerce](https://github.com/woocommerce/woocommerce)?

ilyasfoo commented:
Apologies for having to close this PR now, @zmaglica. Since WCAdmin is merged as per @chihsuan mentioned, code changes can no longer be applied to this repository. Appreciate the effort you've put into this PR, and we look forward for you to migrate the changes to WooCommerce if you have the time for it. Cheers!
